README.md
# completion-cli

[![NPM version][npm-image]][npm-url]
[![build status][ci-image]][ci-url]
[![Test coverage][codecov-image]][codecov-url]
[![npm download][download-image]][download-url]

`completion-cli` is a NodeJS command-line tool to complete text or text files. It uses the OpenAI completions endpoint.

In this context, _completion_ means summarize, continue the "flow", explaining text, etc.

This cli tool was done for personal use, but feel free to use it. When passing whole text or html files, it creates a ReadableStream and sends it by 15Mb chunks to server.

**Requirements**

- [NodeJS v19+](https://nodejs.org/en/download/current/) as `fetch` is used to query the server.

**Security**

- The `apiKey` is only known to you and the OpenAI servers.

## Installation

`npx completion-cli --help`

will give you the options.

It is also possible to download and run it if you have `ts-node`, using `--esm` option:

```shell
ts-node --esm src/index.ts --filePath ../input.txt -k ${OPENAI_API_KEY} -j config.json -t 150
```
## Usage Examples

<details>

<summary> See Usage Examples </summary>

- Simplest one. Request creative reply of 2000 tokens long.

```shell
npx completion-cli -k this5436is546thekey25 -t 2000 -T 1 -p "Explain me the world:"
```

- Send the text from `blogpost.txt` using supported option in the config file.

I tend to use this option almost exclusively.

```shell
npx completion-cli -k this5436is546thekey25 -j ./path/to/config.json -f ./path/to/blogpost.txt
```

Example of a [json config here](./test/jsonConfig.json)

</details>
<br/>

## Options

Pass either cli flags or a JSON-config path. The CLI and JSON options are combined, CLI overwrites JSON. But only one of `--prompt` or `--filePath` or `json.prompt` is allowed.

<details>
<summary> Cli Supported Options </summary>

- `--help`
- `--apiKey` or `-k`, your openAI api key, generated at [OpenAI Official Site](https://beta.openai.com/). **Example**: `completion-cli -k this5436is546thekey25`
- `--filePath` or `-f`, the relative path to the text file. **Example**: `completion-cli -f ./path/to/blogpost.txt`
- `--jsonConfig` or `-j`, path to a json file with any of the [Open AI Supported Options](#supported-onot both, and any of these overwrites ptions). **Example**: `completion-cli -j ./path/to/config.json`

The CLI only supports a small subset of [OpenAI options][1]

- `--prompt` or `-p`: one way to pass text. **Example**: `completion-cli -p Explain me the world: `. The length limit is 15000 bytes (ascii characters.)
- `--model` or `-m`, the default model used is _text-davinci-003_. Other flavours are: _text-curie-001_, _text-babbage-001_. Default `text-davinci-003`
- `--tokens` or `-t` the expected response length in tokens (max is 4096, about 1700 english words.)`
- `--temperature` or `-T` (capital **T**). How creative the model is. Float between 0 and 1. Default 0.
- `--echo` or `-e`. Explicit `boolean`. Indicates whether to concatenate the response to the prompt or not. You need to specify `--echo [true, false]`

</details>

<details>

<br/>

All the values in the object are optional. For explanations and default values check [the schema file](./OpenAIOptions).

<summary>JSON Options</summary>

```js
{
  "apiKey": string;
  "model": string;
  "prompt": string;
  "suffix": string;
  "max_tokens": number;
  "temperature": number;
  "top_p": number;
  "echo": boolean;
  "stop": string | string[];
  "presence_penalty": number;
  "frequency_penalty": number;
  "prePromptString": string;
  "postPromptString": string;
}
```

</details>

<br/>

## Bugs

The package is new and there will be errors. Please file issues or PRs.

## ToDos

- [ ] Support different text encoding for filePath (streams have option for this)
- [x] Support a verbose option for extra logs.
- [ ] Improve boolean flags (atm you have to specify `--flagName true/false`).
- [ ] Support multiple completions from json configuration using the option `n`.
- [ ] Support for `stream` option.
- [x] Support html (cleaning up the tags.). This may be done during stream.
- [ ] Support pdf.
- [ ] Support both from urls as well.

The code is being thought such that it is not too difficult to adapt it to the browser.

## License

[MIT](./LICENSE)

There are no restrictions and you can do with the code what you prefer.

[1]: https://beta.openai.com/docs/api-reference/completions/create
[npm-image]: https://img.shields.io/npm/v/completion-cli.svg
[npm-url]: https://www.npmjs.com/package/completion-cli
[ci-image]: https://github.com/santimirandarp/completion-cli/workflows/Node.js%20CI/badge.svg?branch=main
[ci-url]: https://github.com/santimirandarp/completion-cli/actions?query=workflow%3A%22Node.js+CI%22
[codecov-image]: https://img.shields.io/codecov/c/github/santimirandarp/completion-cli.svg
[codecov-url]: https://codecov.io/gh/santimirandarp/completion-cli
[download-image]: https://img.shields.io/npm/dm/completion-cli.svg
[download-url]: https://www.npmjs.com/package/completion-cli


CHANGELOG.md
# Changelog

## [1.1.0](https://github.com/santimirandarp/completion-cli/compare/v1.0.4...v1.1.0) (2023-01-17)


### Features

* basic html support ([cccfd35](https://github.com/santimirandarp/completion-cli/commit/cccfd3580f030dcfa659e57e2cee39f3c03133fc))

## [1.0.4](https://github.com/santimirandarp/completion-cli/compare/v1.0.3...v1.0.4) (2023-01-16)


### Bug Fixes

* README ([380c053](https://github.com/santimirandarp/completion-cli/commit/380c0530aa4fdc57f987273048f32c3497a45a32))
* rollback to 15Mb ([2966d26](https://github.com/santimirandarp/completion-cli/commit/2966d26f59ff8736c362a4c3435741eed11ed0da))

## [1.0.3](https://github.com/santimirandarp/completion-cli/compare/v1.0.2...v1.0.3) (2023-01-06)


### Bug Fixes

* use only bin for the binary ([fbc9bfc](https://github.com/santimirandarp/completion-cli/commit/fbc9bfc2fb573e50e5f5d910dc9564e0124b78ab))

## [1.0.2](https://github.com/santimirandarp/completion-cli/compare/v1.0.1...v1.0.2) (2023-01-06)


### Bug Fixes

* add binary key to package json ([5abe74c](https://github.com/santimirandarp/completion-cli/commit/5abe74c12b41f8971853bbcf1e3b993598c1383f))

## [1.0.1](https://github.com/santimirandarp/completion-cli/compare/v1.0.0...v1.0.1) (2023-01-06)


### Bug Fixes

* dummy fix, bump version for npm ([5fedb79](https://github.com/santimirandarp/completion-cli/commit/5fedb79fa17edcd8fbc596a7fba39dfdd848bb64))

## 1.0.0 (2023-01-06)


### Features

* completion-cli ([4cbdd88](https://github.com/santimirandarp/completion-cli/commit/4cbdd88358509411f5f096f6959183312cce3427))


src/cliOptions.ts
import meow from 'meow';
export const cliOptions = meow(
  `
  Options
    - --apiKey or -k, expects your openai api key, generated at [openai official site](https://beta.openai.com/). **Example**: completion-cli -k this5436is546thekey25
    - --filePath or -f, the relative path to the text file. **Example**: completion-cli -f ./path/to/blogpost.txt
    - --jsonConfig or -j, path to a json file with any of the [Open AI Supported Options](#supported-options). **Example**: completion-cli -j ./path/to/config.json
    - --verbose or -v, gives you some feedback as it executes. Default false.

    The CLI only supports a small subset of [OpenAI options](https://beta.openai.com/docs/api-reference/completions/create)

    - --prompt or -p: one way to pass text. **Example**: completion-cli -p Explain me the world: 
    - --model or -m, the default model used is _text-davinci-003_. Other flavours are: _text-curie-001_, _text-babbage-001_. Default text-davinci-003
    - --tokens or -t the expected response length in tokens (max is 4096, about 1700 english words.)
    - --temperature or -T (capital **T**). How creative the model is. Float between 0 and 1. Default 0.
    - --echo or -e. Whether to echo the prompt back with the response. Default false.
    - --help or -h, show this help message

  Examples
    $ completion-cli -k this5436is546thekey25 -j ./path/to/config.json -f ./path/to/blogpost.txt
    $ completion-cli -k this5436is546thekey25 -t 2000 -T 1 -p Explain me the world:
  `,
  {
    importMeta: import.meta,
    flags: {
      apiKey: {
        type: 'string',
        alias: 'k',
      },
      filePath: {
        type: 'string',
        alias: 'f',
      },
      jsonConfig: {
        type: 'string',
        alias: 'j',
      },
      prompt: {
        type: 'string',
        alias: 'p',
      },
      model: {
        type: 'string',
        alias: 'm',
      },
      tokens: {
        type: 'number',
        alias: 't',
      },
      temperature: {
        type: 'number',
        alias: 'T',
      },
      echo: {
        type: 'boolean',
        alias: 'e',
      },
      verbose: {
        type: 'boolean',
        alias: 'v',
      },
    },
  },
).flags;

export type CliOptions = typeof cliOptions;
